crates/workspace/src/pane.rs

```diff
@@ -797,6 +797,34 @@ impl Pane {
         destination_index: Option<usize>,
         cx: &mut ViewContext<Self>,
     ) {
+        let max_tabs = WorkspaceSettings::get_global(cx).max_tabs;
+        if let Some(max_tabs) = max_tabs {
+            // Reduce over the activation history to get every dirty items up to max_tabs
+            // count.
+            let mut index_list = Vec::new();
+            let mut items_len = self.items_len();
+            for entry in self.activation_history.iter() {
+                if items_len < max_tabs.get() {
+                    break;
+                }
+                let index_to_remove = self.items.iter().enumerate().find_map(|(index, item)| {
+                    (!item.is_dirty(cx) && item.item_id() == entry.entity_id).then_some(index)
+                });
+                if let Some(index_to_remove) = index_to_remove {
+                    index_list.push(index_to_remove);
+                    items_len -= 1;
+                }
```

Comment:
Here's a diff that would actually avoid the n² complexity. But TBH I'm not sure it is more efficient as we have to build a hash every time. Since we're talking small numbers I already feel like the first optimisation was overkill... Anyway, both are fine to me.
```suggestion
            let mut indexes: HashMap<EntityId, usize> = HashMap::default();
            for (index, item) in self.items.iter().enumerate() {
                indexes.insert(item.item_id(), index);
            }
            for entry in self.activation_history.iter() {
                if items_len < max_tabs.get() {
                    break;
                }
                if !indexes.contains_key(&entry.entity_id) {
                    continue;
                }
                let index = indexes[&entry.entity_id];
                if self.items[index].is_dirty(cx) {
                    continue;
                }

                index_list.push(index);
                items_len -= 1;
```

Reply:
I think the small numbers here mean it should be fine :)